_posts/2019/2019-07-25-how-to-generate-mac-addresses.md

```diff
@@ -0,0 +1,113 @@
+The registration is made by IEEE and costs around 1000$ - 3000$.
+IEEE also publishes a list of already existing [OUIs](http://standards-oui.ieee.org/oui.txt){:target="_blank"}.
+
+## Mac address prefix generation
+
+To simply generate a MAC address prefix, we can have the following options for the first byte:
+
+```
+bbbbbb10 , bbbbbb01, bbbbbb00, bbbbbb11
+```
+where b ∊ {0,1} can be set at random.
+
+The first one (10) is a locally administered unicast address, and it's the most common use case if you need a local MAC address.
+The second one (01) is a globally unique multicast address, and very rare, since multicast addresses are not used very often.
+Our third one (00) is a very typical representative of its species: the global unique unicast address.
+This one is the most common OUI used by almost any hardware producing company (or even software-related when it comes to virtualization). 
+Almost everyone who needs a OUI uses this pattern.
+Last but not least (11) is the locally administered multicast, and also a rare one.
+
+MAC addresses are commonly written in hex and not binary representation, so let's take a look how our candidates look in hex.
+First thing, we can ignore the last 4 bits since they do not affect the b0 and b1 bit in hex. 
+Also b2 and b3 can be 0 or 1 since they do not matter, but affect the hex representation.
+
+We can have the following combination:
+
+```
+ b3    b2    b1 b0    partly decimal    decimal combination   hex representation
+[0|1] [0|1]  0  0     - b3,b2 + 0         - 0,4,8,12            - 0,4,8,C
+[0|1] [0|1]  0  1     - b3,b2 + 1         - 1,5,9,13            - 1,5,9,D
+[0|1] [0|1]  1  0     - b3,b2 + 2         - 2,6,10,14           - 2,6,A,E
+[0|1] [0|1]  1  1     - b3,b2 + 3         - 3,7,11,15           - 3,7,B,F
+```
+Our first byte can have the following hex representations:
+
+```
+[0..F][0|4|8|C] - global unique unicast
+[0..F][1|5|9|D] - global unique multicast
+[0..F][2|6|A|E] - locally administered unicast
+[0..F][3|7|B|F] - locally administered multicast
+```
+
+We now have 2 bytes left, and since they do not have special bits they can be randomly chosen.
+Let's take a look at the following example MAC prefix `EA:A2:B7`.
+`EA` represents our first byte in hex so `A2` and `B7` can be ignored since they are just random.
+The `E` also is just random so the only interesting part is the `A`. 
+When we take a look at our pattern above, we see that `A` is part of the locally administered unicast addresses. 
+`EA:A2:B7` is a valid local unicast MAC address prefix.
+
+The last 3 bytes, let's call them the suffix, can also be generated randomly or simply iteratively.
+An example could be `D4:FC:AE`, and using our example prefix, it would give us the following MAC address `EA:A2:B7:D4:FC:AE`. 
+I've put all this into a [golang](https://golang.org/){:target="_blank"} libary called [mac-gen-go](https://github.com/cseeger-epages/mac-gen-go){:target="_blank"}. 
```

Comment:
Can you  make this a public repo on the epages organization?

Reply:
no i dont think that makes sense since i'am the maintainer and the projekt is way to small to create the overhead of managing the project outside of my own git.

Reply:
The reason for that question was that this link won't be working once you e.g. change your username. We might need to find another solution for this later on but I guess we can still publish it like this then on Thursday :)

Reply:
actually it should work even after changing the username since github will create a redirect until the username is used again

Reply:
https://help.github.com/en/articles/changing-your-github-username#repository-references